src/analyzing/AST_to_IL.ml

```diff
@@ -859,6 +859,41 @@ and record env ((_tok, origfields, _) as record_def) =
                _;
              } ->
              Some (Spread (expr env e))
+         | G.F
+             {
+               s =
+                 G.ExprStmt
+                   ( ({
+                        e =
+                          Call
+                            ( { e = N (Id (id, _)); _ },
+                              (_, [ Arg { e = Record fields; _ } ], _) );
+                        _;
+                      } as prior_expr),
+                     _ );
+               _;
+             }
+           when is_hcl env.lang ->
+             (* This is an inner block of the form
+                someblockhere {
+                  s {
+                    <args>
+                  }
+                }
+
+                We want this to be understood as a record of { <args> } being bound to
+                the name `s`.
+
+                So we just translate it to a field defining `s = <record>`.
```

Comment:
Could we do this transformation at parsing time in Parse_hcl_tree_sitter.ml?

Reply:
i think in principle, no, because the main reason why we don't already is because some blocks have more than one "parameter":
```
s "parameter1" "parameter2" {
  <stuff>
}
```
this we can't really nicely translate to a `DefStmt`, because we want the matchability. the blocks I'm concerned with are of the form that they don't take in any extra parameters, however:
```
s {
  <stuff>
}
```

so I feel like this could in fact be done at parse time, by translating it to a `DefStmt` defining a record

it would be a digression from the general case of a block, however. you'd see that some blocks would get translated to `DefStmt` and some would become `ExprStmt (Call ...)`. this lack of homogeneity could cause some manner of confusion. what do you think?

Reply:
Could we transform the s "parameter1" "parameter2" { } in a DefStmt FuncDef?
If it's really defining a function, then let's encode it as a function?
and then if it has no parameter, it's a variable and we can use DefStmt VarDef, looks simple no?

Reply:
It looks simple, but unfortunately it's not. I said "parameter", but I didn't mean parameter in the sense of a function, I just meant parameter to a block. An instance of this block may be:

resource "type" "name" {
  arg1 = 1
  arg2 = "hi"
}

where "type" is the type the resource ascribes to, "name" is the name of the resource, and the arguments are the parameters used to configure the resource. It may also look like

settings {
  arg1 = true
}

in the case of a nested block. In neither case is either block truly a "function", and it was my predecessor's choice to keep them as Call's, which is the convention I've kept up with. My question is whether the relative cost of special casing the blocks that look like `settings`, at AST_generic time, is less confusing than special casing it at AST_to_IL time.

Reply:
It's also worth noting that these are all Terraform specific notions, and that we claim to support "HCL", which is the more general language surrounding it. If we special case, it will go against the general scheme of HCL.

Reply:
@aryx any thoughts?

Reply:
@aryx  I am a little hesitant to do this, because I would prefer that we translate things which are of the same syntactic class (blocks in HCL) to the same kind of construct. I feel like I would forget all the time if some blocks translated to `DefStmt`s, and some translated to `Call`s, nevermind the fact that something like
```
lifecycle {
  settings1 = 2
}
```
is not really "defining" a name of "lifecycle", but it is instead just a parameter that happens to be labelled "block", that contains some settings.

And indeed, none of these are really functions in any sense.

Right now, there is a nice invariant in `Terraform.ml` that any `DefStmt`s are strictly introduced via `transform_ast_for_analyzing`, or otherwise our pre-analysis Terraform-specific processing steps. This is nice because they truly are from `variable` blocks which do define names, so I would prefer to keep them as the only sources of truth for things which define names, because otherwise our logic which is intended to interact strictly with those definitions will also coincide with these blocks here, even though they are not definitions. Does that make sense?

Reply:
Looking at `Parse_hcl_tree_sitter`, I think it would be better to translate this as a `DefStmt` as Yoann suggests. Although I would always translate it into a `VarDef`/`FieldDefColon` like `<block-id> = (<block-params>, <block-body>)`. So e.g. this below could be represented as `x = (["parameter1", "parameter2"], <stuff>)`.

```hcl
s "parameter1" "parameter2" {
  <stuff>
}
```

Reply:
The thing is that the Terraform translation is arbitrary, we use a `Call` as we could use something else. So it's better to change the `Call` to a `DefStmt` or whatever is needed, rather than keep the `Call` for no particular reason and then add extra code in `AST_to_IL` to cope with that.